### ISSUE TITLE
[ENH] Allow changing the meaning of `int`?

### ISSUE THREAD

**Issue by seberg** [cython/cython#5792]
### Is your feature request related to a problem? Please describe.

In NumPy, we are trying to change the default integer to be `Py_ssize_t` (or equivalent), because `long` is really not very nice on windows 64bit at least.  (yes, there is an argument to be made to always use 64bit numbers, but...)

Since `int` is also `long` in Cython for the same historic reasons, I wonder if it would make sense to have a switch that makes it be `Py_ssize_t` instead.
This would mean that in the long term, the NumPy default integer and Cython can align again (if you know you are running NumPy >=2.0).

I think it would also reduce the number of latent bugs around large integers on windows, while updating code that needs to realign with NumPy, I noticed that in many cases changing the meaning of `int` would make sense, even if would not actually help much with the NumPy transition problem (since it doesn't matter w.r.t. to supporting both NumPy 1.x and 2.x).

One subtlety is that I think `int` conversion from Python currently converts more agressively than other integers (using `int()`/`tp_int` rather than `operator.index()`/`nb_index`  (although I am not sure that is actually intentional).

**Comment by da-woods** [comment]
I think the conversion using `index` (used for `Py_ssize_t`) is probably more restrictive than for `int` - I know we've seen at least two bug reports for this and the Numpy bindings that we've directed to you.

One thing that may be useful - we only apply that rule to `Py_ssize_t` used literally. So if you so:

```
cdef extern from *:
    """
    typedef Py_ssize_t my_big_int
    """
    ctypedef int my_big_int
```

Then in C the typedef is for `Py_ssize_t`. However in Cython it's just "an arbitrary int type", and things like overflow and signedness are left to the C compiler. And Cython doesn't apply its special rules for `Py_ssize_t` (since it doesn't know the underlying C definition).

I know that's not quite what you're asking for, but it may be useful related information.

**Comment by seberg** [comment]
> I know we've seen at least two bug reports for this and the Numpy bindings that we've directed to you.

Right, it seems a bit like it might have changed at some point?  If this was a while ago, I honestly don't care too much.  In general I prefer the use of `index` and using `int(value)` explicitly when it is not desired.

TBH, I would be happy if such a switch is possible, also without the special handling  (because it is questionable also, so opting in to both at once seems perfectly fine).

**Comment by seberg** [comment]
Sorry, read mostly the discussion about index and int.  So yeah, I think your above solution seems like a very viable thing.
Maybe including it as a switch is could be neat because it seems a bit more "obvious"?  (and of course if the `int()` usage can be aligned for backcompat that would be a bonus)

**Comment by seberg** [comment]
Aha OK...  So the change for cython 2 to cython 3, is exactly what you said, and I was just too confused to really grock it.

The Cython 3 numpy defines has:
```
cdef extern from "Python.h":
    ctypedef Py_ssize_t Py_intptr_t

cdef extern from "numpy/arrayobject.h":
    ctypedef Py_intptr_t npy_intp
    ctypedef size_t npy_uintp
```
while the Cython 2 has:
```
cdef extern from "Python.h":
    ctypedef int Py_intptr_t
    bint PyObject_TypeCheck(object obj, PyTypeObject* type)

cdef extern from "numpy/arrayobject.h":
    ctypedef Py_intptr_t npy_intp
    ctypedef size_t npy_uintp
```
Cython deals just fine with the fact that the size of the actual type might be larger if you `ctypedef int myint`.  Changing it to `Py_ssize_t` there should (presumably) do nothing besides changing how the conversion happens, right?

---

So, after removing that confusion, everything seems already possible, it is just that `Py_ssize_t` is special w.r.t. to conversion, which is surprising!

**This makes my ask different:**  It would be nice to have a file wide pre-processor directive, to make any non-explicit integer conversion go through the "safe" path that is currently used exclusively for `Py_ssize_t` (which means going via `PyIndex_*` if the object is not a Python integer already).

In principle of course, it would be even cooler if we could add a `"deprecate"` option so that libraries can transition to it (just like NumPy did very long ago).

---

As for NumPy, it seems that in order to avoid odd downstream fallout, we should probably not define `intp` to `Py_ssize_t` even if it has the same size in practice.

**Comment by seberg** [comment]
OK, I think the fix for Cython 3 and NumPy is to change the definition to:
```
cdef ssize_t npy_intp  # ssize_t is non-standard but Cython "knows" it.
```
Now, I don't think that gets promotion within Cython quite right, but that is a subtle Cython bug, and it will not be any worse than the previous `cdef int npy_intp`.

I found `rank_to_type_name` in `PyrexTypes.py` and maybe one could improve the definitions there to be more complete with respect to integer ranks, but it gets tricky, especially when mixing signed/unsigned...
In the end, users should probably define the type strictly normally, anyway...

**Comment by seberg** [comment]
Final iteration here: https://github.com/numpy/numpy/pull/25094 although backported it by just having the Cython 3 be the same definition as the old Cython 2 one.

`ssize_t` actually also has special rules, it uses `PyLong_AsSize_t`, so `signed long` seemed like the best bet.
